### PR TITLE
#172 Resource detectors for Docker

### PR DESCRIPTION
**Pull request by swetharavichandrancisco**
- Added Resource detectors for Docker environment to extract container.id
- Added Unit Tests

### CODE REVIEW

**Comment by cijothomas** [comment]
@swetharavichandrancisco could you resolve conflicts and rebase?

File: src/OpenTelemetry.Contrib.Extensions.Docker/OpenTelemetry.Contrib.Extensions.Docker.csproj [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>;net461;netstandard2.0</TargetFrameworks>
+    <Description>OpenTelemetry extensions for Docker</Description>
```
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Extensions.Docker/OpenTelemetry.Contrib.Extensions.Docker.csproj:4]: bit more descriptive, as this is very specifically for Resource detection.

**Comment by utpilla** [comment]
Please remove the word `Contrib` from the file names, folder names, namespaces, etc. We have removed the word `Contrib` for rest of the packages in the repo: #196

File: src/OpenTelemetry.Contrib.Extensions.Docker/Resources/DockerResourceDetector.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+using OpenTelemetry.Resources;
+
+namespace OpenTelemetry.Contrib.Extensions.Docker.Resources
+{
+    /// <summary>
+    /// Resource detector for application running in Docker environment.
+    /// </summary>
+    public class DockerResourceDetector : IResourceDetector
+    {
+        private const string FILEPATH = "/proc/self/cgroup";
```
Comment (utpilla) [inline src/OpenTelemetry.Contrib.Extensions.Docker/Resources/DockerResourceDetector.cs:30]: Which is better to use here? `readonly' or `const`?

File: test/OpenTelemetry.Contrib.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+        [Fact]
+        public void TestExtractContainerId()
+        {
+            var dockerResourceDetector = new DockerResourceDetector();
+            Assert.Equal("da261b882cd310e64ced2e31b938f9289261e4c7d59efb2d651ca2fe802b7764", dockerResourceDetector.ExtractContainerId(DOCKERSAMPLEFILEPATH));
+        }
+
+        [Fact]
+        public void TestGetIdFromLine()
```
Comment (utpilla) [inline test/OpenTelemetry.Contrib.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs:76]: Could we simply test for the different possible test cases using `Resource.Attributes` instead of testing the helper methods?

I think we should mark the methods: `ExtractContainerId` and `GetIdFromLine` as `private`.

File: test/OpenTelemetry.Contrib.Extensions.Docker.Tests/Resources/SampleFile/testcgroup [hunk 4]
```diff
@@ -0,0 +1,14 @@
+14:name=systemd:/docker/da261b882cd310e64ced2e31b938f9289261e4c7d59efb2d651ca2fe802b7764
```
Comment (utpilla) [inline test/OpenTelemetry.Contrib.Extensions.Docker.Tests/Resources/SampleFile/testcgroup:1]: What's the purpose of this file?

**Comment by utpilla** [comment]
Add CHANGELOG.md and update the `component_owners.yml` file for this project.

File: src/OpenTelemetry.Contrib.Extensions.Docker/DockerEventSource.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+// </copyright>
+
+using System;
+using System.Diagnostics.Tracing;
+using System.Globalization;
+using System.Threading;
+
+namespace OpenTelemetry.Contrib.Extensions.Docker
+{
+    [EventSource(Name = "OpenTelemetry-Docker")]
```
Comment (utpilla) [inline src/OpenTelemetry.Contrib.Extensions.Docker/DockerEventSource.cs:24]: Use `OpenTelemetry-Extensions-Docker` as the EventSource name.

File: src/OpenTelemetry.Contrib.Extensions.Docker/DockerEventSource.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
+﻿// <copyright file="DockerEventSource.cs" company="OpenTelemetry Authors">
```
Comment (utpilla) [inline src/OpenTelemetry.Contrib.Extensions.Docker/DockerEventSource.cs:1]: nit: Rename the file to `DockerExtensionsEventSource.cs`

File: src/OpenTelemetry.Contrib.Extensions.Docker/DockerEventSource.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+{
+    [EventSource(Name = "OpenTelemetry-Docker")]
+    internal class DockerEventSource : EventSource
+    {
+        public static DockerEventSource Log = new DockerEventSource();
+
+        [NonEvent]
+        public void ResourceAttributesExtractException(string format, Exception ex)
+        {
+            if (this.IsEnabled(EventLevel.Warning, (EventKeywords)(-1)))
```
Comment (utpilla) [inline src/OpenTelemetry.Contrib.Extensions.Docker/DockerEventSource.cs:32]: Use `EventLevel.Error`?

File: test/OpenTelemetry.Contrib.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        {
+            var dockerResourceDetector = new DockerResourceDetector();
+            Assert.Null(dockerResourceDetector.GetIdFromLine(INVALIDCGROUPLINE));
+            Assert.Equal(CONTAINERIDWITHPREFIXREMOVED, dockerResourceDetector.GetIdFromLine(CGROUPLINEWITHPREFIX));
+            Assert.Equal(CONTAINERIDWITHSUFFIXREMOVED, dockerResourceDetector.GetIdFromLine(CGROUPLINEWITHSUFFIX));
+            Assert.Equal(CONTAINERIDWITHPREFIXANDSUFFIXREMOVED, dockerResourceDetector.GetIdFromLine(CGROUPLINEWITHPREFIXandSUFFIX));
+            Assert.Equal(CONTAINERID, dockerResourceDetector.GetIdFromLine(CGROUPLINE));
+        }
+    }
+}
```
Comment (utpilla) [inline test/OpenTelemetry.Contrib.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs:86]: Add a test case for the case when containerId is not found and it's returned as `null`.

File: src/OpenTelemetry.Extensions.Docker/README.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+```csharp
+using OpenTelemetry;
+using OpenTelemetry.Extensions.Docker;
+
+var tracerProvider = Sdk.CreateTracerProviderBuilder()
+                        // other configurations
+                        .SetResourceBuilder(ResourceBuilder
+                            .CreateDefault()
+                            .AddDockerDetector())
```
Comment (utpilla) [inline src/OpenTelemetry.Extensions.Docker/README.md:27]: Where is the extension method defined?

Comment (utpilla) [inline src/OpenTelemetry.Extensions.Docker/README.md:27]: Name the method `AddDockerResourceDetector`.

File: src/OpenTelemetry.Extensions.Docker/DockerExtensionsEventSource.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace OpenTelemetry.Extensions.Docker
+{
+    [EventSource(Name = "OpenTelemetry-Extensions-Docker")]
+    internal class DockerExtensionsEventSource : EventSource
+    {
+        public static DockerExtensionsEventSource Log = new DockerExtensionsEventSource();
+
+        [NonEvent]
+        public void ResourceAttributesExtractException(string format, Exception ex)
```
Comment (utpilla) [inline src/OpenTelemetry.Extensions.Docker/DockerExtensionsEventSource.cs:30]: nit: Rename the method to `ExtractResourceAttributesException`.

File: test/OpenTelemetry.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+// limitations under the License.
+// </copyright>
+
+using System.IO;
+using System.Linq;
+using OpenTelemetry.Extensions.Docker.Resources;
+using OpenTelemetry.Resources;
+using Xunit;
+
+namespace OpenTelemetry.Extensions.Docker.Tests.Resources
```
Comment (utpilla) [inline test/OpenTelemetry.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs:23]: The namespace should be `OpenTelemetry.Extensions.Docker.Tests`. The test explorer on Visual Studio would use the class name to categorize the different unit tests under the namespace.

File: test/OpenTelemetry.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+{
+    public class TestDockerResourceDetector
```
Comment (utpilla) [inline test/OpenTelemetry.Extensions.Docker.Tests/Resources/TestDockerResourceDetector.cs:25]: Rename this to `DockerResourceDetectorTests`. That's the convention followed in the contrib repo.

File: src/OpenTelemetry.Extensions.Docker/OpenTelemetry.Extensions.Docker.csproj [hunk 13]
```diff
@@ -0,0 +1,9 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>;net461;netstandard2.0</TargetFrameworks>
+    <Description>OpenTelemetry Extensions - Container Resource Detector from Docker environment.</Description>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="OpenTelemetry" Version="1.2.0-rc4" />
```
Comment (utpilla) [inline src/OpenTelemetry.Extensions.Docker/OpenTelemetry.Extensions.Docker.csproj:7]: You could use the latest stable version of OTel SDK: `1.2.0`.

File: test/OpenTelemetry.Extensions.Docker.Tests/OpenTelemetry.Extensions.Docker.Tests.csproj [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    <PackageReference Include="xunit" Version="$(XUnitPkgVer)" />
+    <PackageReference Condition="$([MSBuild]::IsOsPlatform('Windows'))" Include="xunit.runner.visualstudio" Version="$(XUnitRunnerVisualStudioPkgVer)">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
+    </PackageReference>
+    <DotNetCliToolReference Include="dotnet-xunit" Version="$(DotNetXUnitCliVer)" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\src\OpenTelemetry.Extensions.Docker\OpenTelemetry.Extensions.Docker.csproj" />
```
Comment (utpilla) [inline test/OpenTelemetry.Extensions.Docker.Tests/OpenTelemetry.Extensions.Docker.Tests.csproj:20]: Use `$(RepoRoot)` instead of `..\..\`.

**Comment by utpilla** [comment]
@swetharavichandrancisco Thanks for updating the PR. There have been a few updates to the Contributing guidelines for this repo since you initially created this PR. Please add issue template, CI flow, etc. Here is the link to the doc for your reference: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md#contributing-a-new-project.

**Comment by swetharavichandrancisco** [comment]
@utpilla, I have followed the contributing guidelines (from https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md#contributing-a-new-project) for my repo and updated the same as well.

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry.Extensions.Docker/DockerExtensionsEventSource.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        [NonEvent]
+        public void ExtractResourceAttributesException(string format, Exception ex)
+        {
+            if (this.IsEnabled(EventLevel.Error, (EventKeywords)(-1)))
+            {
+                this.FailedToExtractResourceAttributes(format, ToInvariantString(ex));
+            }
+        }
+
+        [Event(1, Message = "Failed to extract resource attributes in '{0}'.", Level = EventLevel.Error)]
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions.Docker/DockerExtensionsEventSource.cs:38]: only 1 parameter?

File: src/OpenTelemetry.Extensions.Docker/OpenTelemetry.Extensions.Docker.csproj [hunk 16]
```diff
@@ -0,0 +1,10 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>;net461;netstandard2.0</TargetFrameworks>
+    <Description>OpenTelemetry Extensions - Container Resource Detector from Docker environment.</Description>
+    <MinVerTagPrefix>Extensions.Docker-</MinVerTagPrefix>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="OpenTelemetry" Version="1.2.0" />
+  </ItemGroup>
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions.Docker/OpenTelemetry.Extensions.Docker.csproj:9]: we can use 1.3.0 now

File: src/OpenTelemetry.Extensions.Docker/OpenTelemetry.Extensions.Docker.csproj [hunk 17]
```diff
@@ -0,0 +1,10 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>;net461;netstandard2.0</TargetFrameworks>
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions.Docker/OpenTelemetry.Extensions.Docker.csproj:3]: I'd suggest net462. net461 is out of support already.

File: src/OpenTelemetry.Extensions.Docker/README.md [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+                        .SetResourceBuilder(ResourceBuilder
+                            .CreateEmpty()
+                            .AddDetector(new DockerResourceDetector()))
+                        .Build();
+```
+
+The resource detectors will record the following metadata based on where
+your application is running:
+
+- **DockerResourceDetector**: container id.
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions.Docker/README.md:34]: would be nice to list the name of the attribute used

File: src/OpenTelemetry.Extensions.Docker/Resources/DockerSemanticConventions.cs [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.Extensions.Docker.Resources
+{
+    internal static class DockerSemanticConventions
+    {
+        public const string AttributeContainerID = "container.id";
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions.Docker/Resources/DockerSemanticConventions.cs:21]: please add a link to the relevant OTel spec/convention

File: test/OpenTelemetry.Extensions.Docker.Tests/OpenTelemetry.Extensions.Docker.Tests.csproj [hunk 20]
```diff
@@ -0,0 +1,23 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <Description>Unit test project for Docker Detector for OpenTelemetry</Description>
+    <TargetFrameworks>netcoreapp3.1</TargetFrameworks>
```
Comment (cijothomas) [inline test/OpenTelemetry.Extensions.Docker.Tests/OpenTelemetry.Extensions.Docker.Tests.csproj:5]: add net6.0 too

File: test/OpenTelemetry.Extensions.Docker.Tests/OpenTelemetry.Extensions.Docker.Tests.csproj [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+    <TargetFrameworks Condition="$(OS) == 'Windows_NT'">$(TargetFrameworks);net461</TargetFrameworks>
```
Comment (cijothomas) [inline test/OpenTelemetry.Extensions.Docker.Tests/OpenTelemetry.Extensions.Docker.Tests.csproj:6]:
```suggestion
    <TargetFrameworks Condition="$(OS) == 'Windows_NT'">$(TargetFrameworks);net462</TargetFrameworks>
```

**Review by cijothomas — APPROVED** [review]
Left few comments. Would be nice to address before 1st release.

**Comment by utpilla** [comment]
> Left few comments. Would be nice to address before 1st release.

@swetharavichandrancisco Please address these comments in a follow-up PR.

**Comment by swetharavichandrancisco** [comment]
@cijothomas @utpilla , I have addressed the review comments of upgrading .NET Framework/OpenTelemetry version in the follow-up PR - https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/432

File: src/OpenTelemetry.Extensions.Docker/Resources/DockerSemanticConventions.cs (hunk 19 above)
Comment (swetharavichandrancisco) [reply src/OpenTelemetry.Extensions.Docker/Resources/DockerSemanticConventions.cs:21]: I have added readme file which provides the usage/description as a part of previous PR - https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Extensions.Docker

I'm not sure if you mean by updating the OpenTelemtry docs - https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/README.md. If I need to update it here, Do I need to provide an example as well?

File: src/OpenTelemetry.Extensions.Docker/DockerExtensionsEventSource.cs (hunk 15 above)
Comment (swetharavichandrancisco) [reply src/OpenTelemetry.Extensions.Docker/DockerExtensionsEventSource.cs:38]: It Prints out the LOG Level with filename and an actual exception. Do you expect more?
